USF-2012: Create Terms and Conditions documentation (#204)
commerce-docs/microsite-commerce-storefront: src/content/docs/dropins/checkout/containers/terms-and-conditions.mdx

---
title: TermsAndConditions container
description: Configure the TermsAndConditions container to manage and display the terms and conditions form during checkout.
---

import Aside from '@components/Aside.astro';
import OptionsTable from '@components/OptionsTable.astro';

> The `TermsAndConditions` container is designed to handle and display a checkbox for the users to agree to the terms and conditions of the sale before the purchase is finalized. The user must check all the agreements to place the order during a guest checkout process, otherwise a validation error will appear when the user clicks on the Place Order button.

Reviewer: ```suggestion
The `TermsAndConditions` container displays a checkbox that users must select to agree to the terms and conditions of the sale before confirming their purchase.

During a guest checkout process, users must check all required agreements before placing an order. If an agreement is unchecked, a validation error appears when the user clicks the **Place Order** button.
```

<Aside type="note" title="TermsAndConditions not displayed for any reason?">
  <li>The `TermsAndConditions` container won't be displayed in case the user is logged into their account, because the customer accepted those agreements previously during the registration process.</li>
  <li>The `TermsAndConditions` container requires a store configuration to be enabled; so it won't be displayed if the component is not properly configured.</li>

>   Please visit [Terms & Conditions setup](/merchant/terms-and-conditions/) to know exactly how to enable and configure the Terms and Conditions feature.

Reviewer: ```suggestion
Visit the [Terms & Conditions setup](/merchant/terms-and-conditions/) documentation for more information on how to enable the Terms and Conditions feature.
```

</Aside>

<Aside type="caution" title="TermsAndConditions requirements">
  In order to use the `TermsAndConditions` container, the **Storefront Compatibility Package (SCP) 4.7.1-beta8** (or higher) module must be installed.
  The **SCP 4.7.1-beta8** added support for retrieving Terms and Conditions configuration setting via the _StoreConfig_ GraphQL query. This setting is required by `TermsAndConditions` container to allow frontend applications to dynamically enable and configure agreements by store-view in checkout page.
</Aside>

## TermsAndConditions configurations

The `TermsAndConditions` container provides the following configuration options:

<OptionsTable
  compact
  options={[
    ['Option', 'Type', 'Req?', 'Description'],
    ['slots', 'object', 'No', 'Object with a list of agreements to be accepted by the guest user.'],
  ]}
/>

> These configuration options are implementing the `TermsAndConditionsProps` interface:

Reviewer: ```suggestion
These configuration options implement the `TermsAndConditionsProps` interface:
```

### TermsAndConditionsProps interface

> The `TermsAndConditions` container receives an object as parameter which implements the `TermsAndConditionsProps` interface with the following properties:

Reviewer: ```suggestion
The `TermsAndConditions` container receives an object as a parameter which implements the `TermsAndConditionsProps` interface with the following properties:
```

```ts
export interface TermsAndConditionsProps {
  slots?: {
    Agreements?: SlotProps<{
      appendAgreement: SlotMethod<{
        name: string;
        mode: AgreementMode;
        text?: string;
        translationId?: string;
      }>;
    }>;
  };
}
```

<li>The `slots` property is an object containing the following properties:</li>

#### Agreements property

> The `Agreements` property is a handler used to render and configure the list of agreements. It provides a context including the method `appendAgreement()` to add a new agreement:

Reviewer: ```suggestion
The `Agreements` property is a handler used to render and configure the list of agreements. It provides a context by including the method `appendAgreement()` to add a new agreement:
```

```ts
export type SlotProps<T = any> = (
  ctx: T & DefaultSlotContext<T>,
  element: HTMLDivElement | null
) => Promise<void> | void;

export type SlotMethod<P = any> = (
  callback: (next: unknown, state: State) => P
) => void;

export enum AgreementMode {
  MANUAL = 'manual',
  AUTO = 'auto',
}

. . .
    Agreements?: SlotProps<{
      appendAgreement: SlotMethod<{
        name: string;
        mode: AgreementMode;
        text?: string;
        translationId?: string;
      }>;
    }>;
. . .
```

<li>The `appendAgreement` configuration is a callback function which accepts the following attributes to configure an agreement:</li>

>   - **`name`**

Reviewer: would it make more sense to put it like this?:

  - **`name`**: The agreement identifier.
  - **`mode`**: Specifies the mode how the checkbox should appear:
       - 'manual': the user should manually check the terms
       - 'auto': the checkbox will appear checked by default
   ... etc

Author: This is exactly how it is displayed on the documentation... please see the following screenshot:

<img width="820" alt="image" src="https://github.com/user-attachments/assets/f60078c5-1075-449f-a44c-d0cfdacbc822" />

---
title: TermsAndConditions container
description: Configure the TermsAndConditions container to manage and display the terms and conditions form during checkout.
---

import Callouts from '@components/Callouts.astro';
import OptionsTable from '@components/OptionsTable.astro';
import { Steps } from '@astrojs/starlight/components';
import Tasks from '@components/Tasks.astro';
import Task from '@components/Task.astro';

> The `TermsAndConditions` container is designed to handle and display a checkbox for each agreement. The user must check all the agreements in order to place the order during a guest checkout process. If any of the agreements is unchecked, a validation error will appear when the user clicks on Place Order button.

Reviewer 2: ```suggestion
The `TermsAndConditions` container is designed to handle and display a checkbox for the customers to agree to the terms and conditions of the sale before the purchase is finalized. The user must check all the agreements to place the order during a guest checkout process. If any of the agreements is unchecked, a validation error will appear when the user clicks on the Place Order button.
```

Reviewer 2: It would be interesting to add something like merchants can add as many agreements as they want on a per store-view basis

Author: I think it's mentioned below, when I explain how to configure it with more details... but I can put a message in the introduction as well. Thxs

Please note the `TermsAndConditions` container won't be displayed in case the user has signed-in, because the customer accepted those agreements previously during the registration process.

It will also not be displayed for a specific store view if the component is not properly enabled on it.

## TermsAndConditions configurations

The `TermsAndConditions` container provides the following configuration options:

<OptionsTable
  compact
  options={[
    ['Option', 'Type', 'Req?', 'Description'],
    ['slots', 'object', 'No', 'Object with a list of agreements to be accepted by the guest user.'],
  ]}
/>

These configuration options are implementing the `TermsAndConditionsProps` interface:

### TermsAndConditionsProps interface

The `TermsAndConditions` container receives an object as parameter which implements the `TermsAndConditionsProps` interface with the following properties:

```ts
export interface TermsAndConditionsProps {
  slots?: {
    Agreements?: SlotProps<{
      appendAgreement: SlotMethod<{

>         mode: AgreementMode;
>         name: string;
>         translationId?: string;
>         text?: string;

Reviewer 2: Reorder alphabetically

```suggestion
        name: string;
        mode: AgreementMode;
        text?: string;
        translationId?: string;

```

Author: It was directly a copy & paste from the code...

      }>;
    }>;
  };
}
```

<li>The `slots` property is an object containing the following properties:</li>

#### Agreements property

The `Agreements` property is a handler used to render and configure the list of agreements. It provides a context including the method `appendAgreement()` to add a new agreement:

```ts
export type SlotProps<T = any> = (
  ctx: T & DefaultSlotContext<T>,
  element: HTMLDivElement | null
) => Promise<void> | void;

export type SlotMethod<P = any> = (
  callback: (next: unknown, state: State) => P
) => void;

export enum AgreementMode {
  MANUAL = 'manual',
  AUTO = 'auto',
}

. . .
    Agreements?: SlotProps<{
      appendAgreement: SlotMethod<{
        mode: AgreementMode;
        name: string;
        translationId?: string;
        text?: string;
      }>;
    }>;
. . .
```

<li>The `appendAgreement` configuration is a callback function which accepts the following attributes to configure an agreement:</li>

  - **`mode`**

>     It defines how the checkbox should be shown:
>     - 'manual' for unchecked; the user should manually check it
>     - 'auto' for checked; the checkbox will appear checked by default

Reviewer 2: ```suggestion
    It specifies how the checkbox should appear:
    - 'manual': the user should manually check the terms.
    - 'auto': the checkbox will appear checked by default
```

1. Finally, into the _"Checkout Options"_ enable the _"Enable Terms and Conditions"_ configuration selecting the _**Yes**_ value for the store view desired.

</Steps>

See [Terms and conditions for checkout](https://experienceleague.adobe.com/en/docs/commerce-admin/stores-sales/point-of-purchase/checkout/terms-and-conditions) documentation to enable the terms and conditions using the Admin Panel for more details.

</Task>

<Task>

### Create the agreement document

The agreements pages will be created **ONLY** using _document-based authoring (DBA)_ within EDS instead of using the Admin Panel, allowing the merchant to create and manage as many agreements pages and content as required for each store view and language.

</Task>

<Task>

### Configure the agreement text label

Since the agreements pages will be created using _DBA_ within EDS, the agreements checkbox text should support different links for each store view, directing them to the appropriate agreements pages.

The merchant has two options in order to configure the text to display next to the checkbox:

<Callouts columnCount="1">

1. **Using EDS only**

>     The merchant has to configure the checkbox label and text for each agreement into the **`placeholders.xlsx`** EDS file. The checkbox label is directly retrieved from the Boilerplate translations for a specific store-view and language.

Reviewer 2: `The merchant has to configure the checkbox label and text for each agreement `
Is there any difference between label and text? I would try to keep the same concept all over the docs

2. **Using Admin Panel**

    The merchant has to configure the checkbox text for each agreement using the Admin Panel. As mentioned before, the Admin Panel allows to create configurations per store view level as following:

</Callouts>

<Steps>

1. Log in to the Admin Panel dashboard.
1. Go to the _"Stores"_ menu and click on the _"SETTINGS" > "Terms and Conditions"_ option.
1. Click the "Add New Condition" button providing the next Terms and Conditions information:

    - **`Condition Name`**
      It's the agreement identifier
    - **`Status`**
      Attribute to indicate if it's available or not:
      - 'Enabled'; the agreement will be retrieved by the GraphQL API
      - 'Disabled'; the agreement won't be retrieved by the GraphQL API
    - **`Applied`**
      It defines the mode how the checkbox should be shown:
      - 'Manually' for unchecked; the user should manually check it
      - 'Automatically' for checked; the checkbox will appear checked by default
    - **`Store View`**
      It defines the scope at store level for this agreement
    - **`Checkbox Text`**
      It contains directly the text to show; it accepts HTML with links to a specific page in EDS

>     Please note the following attributes have not been mentioned and won't be provided by the merchant because they will be handled in EDS:

Reviewer 2: I would try to explain here that the following attributes **are provided**, but they are not utilized in the current solution. Merchant could work on their own solution if they want to use the following attributes

  
      - **`Content`**
        As described above, the content for the agreements pages will be created **ONLY** using EDS
      - **`Show Content as`** (Text, HTML)
        Not really needed because EDS pages accepts HTML code
      - **`Content Height (css)`**

    These entries will be used in order to render the `TermsAndConditions` container inside of the **commerce-checkout.js** block file. Some examples will be provided below.

    See [Terms and conditions for checkout](https://experienceleague.adobe.com/en/docs/commerce-admin/stores-sales/point-of-purchase/checkout/terms-and-conditions) documentation to configure the terms and conditions using the Admin Panel for more details.

</Steps>

</Task>

</Tasks>

## Example 1: Render a custom agreement

> The following example renders the `TermsAndConditions` container on a checkout page, displaying a custom agreement providing directly the text to show with the link to the EDS page, in the element with the class `.checkout__terms-and-conditions`:

Reviewer 2: ```suggestion
The following example renders the TermsAndConditions container on the checkout page, displaying a custom agreement that directly includes the text to show along with the link to the EDS page, within the element having the class `.checkout__terms-and-conditions`:
```

```ts
// Checkout Dropin
import TermsAndConditions from '@dropins/storefront-checkout/containers/TermsAndConditions.js';
import { render as CheckoutProvider } from '@dropins/storefront-checkout/render.js';

const $termsAndConditions = checkoutFragment.querySelector(
  '.checkout__terms-and-conditions',
);

CheckoutProvider.render(TermsAndConditions, {
  slots: {
    Agreements: (ctx) => {
      ctx.appendAgreement(() => ({
        name: 'custom',
        mode: 'auto',

>         text: 'Custom terms and conditions <b><a href="/terms-and-conditions">Terms & Conditions</a></b>.',

Reviewer 2: - Links will have their own CSS so no need to add a bold tag. 
- Remember to add `target='_blank'` to the links in the examples so the user will be redirected to the page on a new tab.

```suggestion
        text: 'Custom terms and conditions <a href="/terms-and-conditions" target="_blank">Terms & Conditions</a>.',
```

Author: The bold tag was just an example to show that the HTML tags are accepted.

      }));
    },
  },
})($termsAndConditions),
```

> ## Example 2: Render the available agreements labels configured in EDS

Reviewer 2: ```suggestion
## Example 2: Render the available agreements using the translations configured in EDS
```

> The following example renders the `TermsAndConditions` container on a checkout page, displaying the available agreements configured into the **`placeholders.xlsx`** EDS file, in the element with the class `.checkout__terms-and-conditions`:

Reviewer 2: ```suggestion
The following example renders the `TermsAndConditions` container on the checkout page, displaying the configured agreements and taking the text from the translations in the **`placeholders.xlsx`** file, within the element with the class `.checkout__terms-and-conditions`:
```

```ts
// Checkout Dropin
import TermsAndConditions from '@dropins/storefront-checkout/containers/TermsAndConditions.js';
import { render as CheckoutProvider } from '@dropins/storefront-checkout/render.js';

const $termsAndConditions = checkoutFragment.querySelector(
  '.checkout__terms-and-conditions',
);

CheckoutProvider.render(TermsAndConditions, {
  slots: {
    Agreements: (ctx) => {
      ctx.appendAgreement(() => ({
        name: 'default',
        mode: 'auto',
        translationId: 'Checkout.TermsAndConditions.label',
      }));
      ctx.appendAgreement(() => ({
        name: 'terms',
        mode: 'manual',
        translationId: 'Checkout.TermsAndConditions.terms_label',
      }));
      ctx.appendAgreement(() => ({
        name: 'privacy',
        mode: 'auto',
        translationId: 'Checkout.TermsAndConditions.privacy_label',
      }));
    },
  },
})($termsAndConditions),
```

## Example 3: Render the available agreements configured in the Admin Panel

> The following example renders the `TermsAndConditions` container on a checkout page, displaying the available agreements configured in the Admin Panel, in the element with the class `.checkout__terms-and-conditions`:

Reviewer 2: I don't think this sentence adds any value to the documentation.
`in the element with the class .checkout__terms-and-conditions`

Author: Just a copy & paste from other drop-ins... but what I mean is that the agreements will be rendered on the element with classname `.checkout__terms-and-conditions`

---
title: TermsAndConditions container
description: Configure the TermsAndConditions container to manage and display the terms and conditions form during checkout.
---

import Callouts from '@components/Callouts.astro';
import OptionsTable from '@components/OptionsTable.astro';
import { Steps } from '@astrojs/starlight/components';
import Tasks from '@components/Tasks.astro';
import Task from '@components/Task.astro';

The `TermsAndConditions` container is designed to handle and display a checkbox for the customers to agree to the terms and conditions of the sale before the purchase is finalized. The user must check all the agreements to place the order during a guest checkout process. If any of the agreements is unchecked, a validation error will appear when the user clicks on the Place Order button.

Please note the `TermsAndConditions` container won't be displayed in case the user has signed-in, because the customer accepted those agreements previously during the registration process.

It will also not be displayed for a specific store view if the component is not properly enabled on it.

> ## TermsAndConditions configurations

Reviewer 2: Should we mention somewhere that the SCP (Storefront Compatibility Package) package version required for the T&C is version `4.7.1-beta8` or higher?

Author: Done.

    ['slots', 'object', 'No', 'Object with a list of agreements to be accepted by the guest user.'],
  ]}
/>

These configuration options are implementing the `TermsAndConditionsProps` interface:

### TermsAndConditionsProps interface

The `TermsAndConditions` container receives an object as parameter which implements the `TermsAndConditionsProps` interface with the following properties:

```ts
export interface TermsAndConditionsProps {
  slots?: {
    Agreements?: SlotProps<{
      appendAgreement: SlotMethod<{
        name: string;
        mode: AgreementMode;
        text?: string;
        translationId?: string;
      }>;
    }>;
  };
}
```

<li>The `slots` property is an object containing the following properties:</li>

#### Agreements property

The `Agreements` property is a handler used to render and configure the list of agreements. It provides a context including the method `appendAgreement()` to add a new agreement:

```ts
export type SlotProps<T = any> = (
  ctx: T & DefaultSlotContext<T>,
  element: HTMLDivElement | null
) => Promise<void> | void;

export type SlotMethod<P = any> = (
  callback: (next: unknown, state: State) => P

> ) => void;

Reviewer 2: I'm not sure if these types add any value to these docs

Author: I prefer to put all the types used

</Task>

</Tasks>

## Example 1: Render a custom agreement

The following example renders the `TermsAndConditions` container on the checkout page, displaying a custom agreement that directly includes the label to show along with the link to the EDS page, within the element having the class `.checkout__terms-and-conditions`:

```ts
// Checkout Dropin
import TermsAndConditions from '@dropins/storefront-checkout/containers/TermsAndConditions.js';
import { render as CheckoutProvider } from '@dropins/storefront-checkout/render.js';

const $termsAndConditions = checkoutFragment.querySelector(
  '.checkout__terms-and-conditions',
);

CheckoutProvider.render(TermsAndConditions, {
  slots: {
    Agreements: (ctx) => {
      ctx.appendAgreement(() => ({
        name: 'custom',
        mode: 'auto',
        text: 'Custom terms and conditions <a href="/terms-and-conditions" target="_blank">Terms & Conditions</a>.',
      }));
    },
  },
})($termsAndConditions),
```

> ## Example 2: Render the available agreements using the translations configured in EDS

Reviewer 2: ```suggestion
## Example 2: Render three different agreements using the translations configured in EDS
```

Reviewer 2: I am adding this proposal since there are no available agreements in this scenario, the merchant just adds as many agreements as they wish

> The following example renders the `TermsAndConditions` container on the checkout page, displaying the configured agreements and taking the label from the translations in the **`placeholders.xlsx`** EDS file, within the element with the class `.checkout__terms-and-conditions`:

Reviewer 2: ```suggestion
The following example renders the `TermsAndConditions` container on the checkout page, displaying three different agreements and taking the label from the translations in the **`placeholders.xlsx`** EDS file:
```
